ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/SearchTransformer.java

```diff
@@ -128,8 +132,23 @@ public Shuttle(final RexBuilder rexBuilder) {
         } else {
           return call;
         }
+      case IS_NULL:
+      case IS_NOT_NULL:
+      case CASE:
+      case COALESCE:
+        // Everything below must be transformed with the UNKNOWN handler
```

Comment:
Can you elaborate on why you can't use the outermost handler here?

In the Jira ticket you seem towards the opposite direction, that is taking into account that in `WHERE` clauses filters obey to the `unknownAsFalse` semantics.

I haven't reviewed the Calcite upgrade which introduced some of the code you touch here, so maybe it should be clear from the context.

Reply:
There are some operators with special semantics that we need to retain the 3-valued logic even if we are inside the `WHERE` clause. 

Example: `COALESCE(SEARCH(...), true)`
1. If SEARCH -> true then COALESCE -> true
2. If SEARCH -> false then COALESCE -> false
3. If SEARCH -> null then COALESCE -> true

If we consider `unknownAsFalse` inside the COALESCE then we will incorrectly drop the `IS NOT NULL` predicate during the expansion and the COALESCE will fall into the third case returning `true` instead of `false`.

Basically, if the SEARCH is below/inside an operator that is sensitive to null values we cannot safely perform the simplification.

Note that `RexSimplify` also changes the default unknown semantics for certain operators (see `simplifyIs2`, `simplifyCoalesce`, etc.).

Reply:
I suspected something similar, thanks for the clarification.

Since this is a rather deep technical detail that most people not familiar with CBO will have problems with, I'd suggest to have a more explicit comment like "// These operator kinds require the UNKNOWN handler to correctly handle NULL values" or something similar.

I know it's still a draft PR but for the rest LGTM already, feel free to ping me for a review if needed.

Reply:
I added more comments and also made the approach a bit more defensive in the sense that if SEARCH happens to be under any operator other than AND/OR we don't perform the simplification.